File: borgbutler-server/src/main/java/de/micromata/borgbutler/server/rest/VersionRest.java
package de.micromata.borgbutler.server.rest;

import de.micromata.borgbutler.json.JsonUtils;
import de.micromata.borgbutler.server.Languages;
import de.micromata.borgbutler.server.Version;
import de.micromata.borgbutler.server.user.UserData;
import org.apache.commons.lang3.StringUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import java.util.Date;
import java.util.Locale;

@Path("/")
public class VersionRest {
    private Logger log = LoggerFactory.getLogger(VersionRest.class);

    /**
     *
     * @param requestContext For detecting the user's client locale.
     * @param prettyPrinter If true then the json output will be in pretty format.
     * @see JsonUtils#toJson(Object, boolean)
     */
    @GET
    @Path("version")
    @Produces(MediaType.APPLICATION_JSON)
    public String getVersion(@Context HttpServletRequest requestContext, @QueryParam("prettyPrinter") boolean prettyPrinter) {
        UserData user = RestUtils.getUser();
        String language = Languages.asString(user.getLocale());
        if (StringUtils.isBlank(language)) {
            Locale locale = requestContext.getLocale();
            language = locale.getLanguage();
        }
        MyVersion version = new MyVersion(language, RestUtils.checkLocalDesktopAvailable(requestContext) == null);
        String json = JsonUtils.toJson(version, prettyPrinter);
        return json;
    }

    public class MyVersion {
        private Version version;
        private String language;
        private boolean localDesktopAvailable;

        private MyVersion(String language, boolean localDesktopAvailable) {
            this.version = Version.getInstance();
            this.language = language;
            this.localDesktopAvailable = localDesktopAvailable;
        }

        public String getAppName() {
            return version.getAppName();
        }

        public String getVersion() {
            return version.getVersion();
        }

        public String getBuildDateUTC() {
            return version.getBuildDateUTC();
        }

        public Date getBuildDate() {
            return version.getBuildDate();
        }

        /**
         * @return Version of the available update, if exist. Otherwise null.
         */
        public String getUpdateVersion() {
            return version.getUpdateVersion();
        }

        public String getLanguage() {
            return language;
        }

        public boolean isLocalDesktopAvailable() {
            return localDesktopAvailable;
        }
    }
}

File: borgbutler-server/src/main/kotlin/de/micromata/borgbutler/server/rest/VersionRest.kt
package de.micromata.borgbutler.server.rest

import de.micromata.borgbutler.json.JsonUtils
import de.micromata.borgbutler.server.Languages
import de.micromata.borgbutler.server.Version
import org.apache.commons.lang3.StringUtils
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import java.util.*
import javax.servlet.http.HttpServletRequest

@RestController
@RequestMapping("/rest")
class VersionRest {
    /**
     *
     * @param request For detecting the user's client locale.
     * @see JsonUtils.toJson
     */
    @GetMapping("version")
    fun getVersion(request: HttpServletRequest): MyVersion {
        val user = RestUtils.getUser()
        var language = Languages.asString(user.getLocale())
        if (StringUtils.isBlank(language)) {
            val locale: Locale = request.locale
            language = locale.language
        }
        return MyVersion(language, RestUtils.checkLocalDesktopAvailable(request) == null)
    }

    inner class MyVersion(val language: String, val localDesktopAvailable: Boolean) {
        private val version: Version = Version.getInstance()
        val appName: String
            get() = version.appName

        fun getVersion(): String {
            return version.version
        }

        val buildDateUTC: String
            get() = version.buildDateUTC
        val buildDate: Date
            get() = version.buildDate

        /**
         * @return Version of the available update, if exist. Otherwise null.
         */
        val updateVersion: String?
            get() = version.updateVersion

    }
}
